### PR TITLE
Added 'names' attribute field parsing in credx

### PR DESCRIPTION
**Pull request by bobozaur**
Addresses #948 .

### CODE REVIEW

File: aries_vcx_core/src/anoncreds/credx_anoncreds.rs
Comment (Patrik-Stas) [inline aries_vcx_core/src/anoncreds/credx_anoncreds.rs:734]: I'd suggest to also make sanity check on name/names first - if none, or both are specific, it's invalid proof request (according to spec). Currently if:
-  both are specified, we don't error but only prioritize `names`
- if none is specified, we fail on line 765 with some error, but I imagine we could do a nicer one (signalling an invalid proof request for reason X).

File: aries_vcx_core/src/anoncreds/credx_anoncreds.rs
Comment (Patrik-Stas) [inline aries_vcx_core/src/anoncreds/credx_anoncreds.rs:748]: Right, it should use the top level `non_revoked` value by default, but get override if specified on specific <requested attribute> level. And it seems like right now it's only looking on the <requested attribute> level, is that right?

File: aries_vcx/src/common/anoncreds.rs
Comment (Patrik-Stas) [inline aries_vcx/src/common/anoncreds.rs:91]: forgotten print

File: aries_vcx_core/src/anoncreds/credx_anoncreds.rs
Comment (Patrik-Stas) [inline aries_vcx_core/src/anoncreds/credx_anoncreds.rs:735]: It feels like a code smell because the `ok_or_else` err handling should be never reached. Perhaps would be cleaner to do a match on tuple of `(requested_val.get("names"), requested_val.get("name"))` and that would also sort out the invalid cases referred to in the comment above.

File: aries_vcx_core/src/anoncreds/credx_anoncreds.rs
Comment (Patrik-Stas) [inline aries_vcx_core/src/anoncreds/credx_anoncreds.rs:763]: Subjective and not directly relevant to this PR, just noticed - the `reft` variable name, would prefer full wording `referent`

File: aries_vcx_core/src/anoncreds/credx_anoncreds.rs
Comment (Patrik-Stas) [inline aries_vcx_core/src/anoncreds/credx_anoncreds.rs:761]: Hmm, I haven't checked how this is done in other implementations, but it seems like something might be missing. If verifier specified `names: ["name", "surname"]` it means that fields from which these fields are selected from are originating from a single credential https://hyperledger.github.io/anoncreds-spec/#presentation-request-example
eg. not mixing "name" from one credential and "surname" from another.

While this code could coincidentally (and in practice currently even likely) work if there's not more alternative credentials in the wallet, it seems like prover is not trying to assure that property.

We don't necessarily have to address this right now, but lets at least create issue to track this.

**Review by Patrik-Stas — CHANGES_REQUESTED** [review]
Left some comments, but happy to see this
https://github.com/hyperledger/aries-vcx/pull/927

now fully passing on top of this PR!

File: aries_vcx/src/common/anoncreds.rs
Comment (bobozaur) [reply aries_vcx/src/common/anoncreds.rs]: Fixed

File: aries_vcx_core/src/anoncreds/credx_anoncreds.rs
Comment (bobozaur) [reply aries_vcx_core/src/anoncreds/credx_anoncreds.rs:763]: Most of this was done by @gmulhearn . I only did the issuer.

**Comment by bobozaur** [comment]
@Patrik-Stas Thanks for taking the time to review this. I want to highlight some things:
- This was "hastily" done to fix the issue. There are far better ways to handle this.
- Most of the implementation, apart from the issuer (which was done by me) was done by @gmulhearn . I'm not all that familiar with all the business logic.
- There's a lot of code smell throughout the entire credx implementation. I would not tweak certain things but rather rewrite it all (it's my motto, after all 😆 ).

Given these things, my approach would be:
- Get this into a suitable working state (if it's not already) and merge it.
- Remove vdrtools anoncreds implementation in favor of credx anoncreds.
- Rewrite ledger trait to add strong typing
- Rewrite credx anoncreds so that:
    - It uses strong typing introduced in the ledger trait refactor and brings its own
    - It is readable and maintainable
    - It is more performant, mainly due to stronger typing and not needing to serialize/deserialize stuff in every single function call.

For instance, the `name` or `names` situation could be handled in an enum type, thus enforcing that a single field of the two are provided. But I'd rather understand all the business logic, get all the prerequisites regarding strong typing done, and then rewrite the credx anoncreds impl rather than patching this single spot in the code pile we have now.

I'd like to hear what you, or the others, think about this roadmap.

File: aries_vcx_core/src/anoncreds/credx_anoncreds.rs
Comment (gmulhearn) [reply aries_vcx_core/src/anoncreds/credx_anoncreds.rs:761]: ye IMO this is the whole point of `names`; requiring that multiple attributes are presented from the same credential. `credentials_json` should only be populated with credentials which have ALL the attributes (names) present

Comment (gmulhearn) [reply aries_vcx_core/src/anoncreds/credx_anoncreds.rs]: Hmm no that doesn't fix it. Consider a request for `names: ["a", "b", "c"]`, and a wallet with `cred1 { a, d }`, `cred2 { b, e }`, and `cred3 { c, f }`.

The tweaked implementation with `all_found` would result in `cred1`, `cred2` and `cred3` being added to `credentials_json` and included. Where really, none of the credentials satisfy the requirement of having all 3 attribute names (a, b, and c), so the credentials for that referent should be empty.

A credential should only be added to the list if it has ALL the `name`/`names` as attributes on the credential, AND it satisfies any restrictions applied to it. 

The ideal way to do that IMO is to rewrite `_get_credentials_for_proof_req_for_attr_name` a bit, such that it only needs to be called once per referent, and it takes in an array of `attr_names`, and the WQL query will search for credentials which have ALL those attributes (currently it just checks for 1 attr_name passed in).

Comment (gmulhearn) [reply aries_vcx_core/src/anoncreds/credx_anoncreds.rs]: Anyway, will leave it up to you guys whether that is scope for this PR or for sometime else; as you've stated, the impl is far from perfect and a rewrite is probs best... the original impl was meant to act as a 'base' for the sake of getting it in there, so it's great that you guys are readdress the impl and improving

**Comment by Patrik-Stas** [comment]
@bobozaur that's a great plan! Sounds good to me 👍

**Review by Patrik-Stas — APPROVED** [review]

**Comment by bobozaur** [comment]
> @bobozaur sounds good, although I am not sure what would you rewrite about the ledger trait (other than adding typing (which would ideally be the case for all of the `aries_vcx_core` components) )
> 
> Anyway, in scope of this PR I am fine with merge

We could still add the associated types. It looks like @gmulhearn has the same concern so that we don't end up with a very rigid interface that only works with credx.

File: aries_vcx_core/src/anoncreds/credx_anoncreds.rs
Comment (bobozaur) [reply aries_vcx_core/src/anoncreds/credx_anoncreds.rs]: Okay I had another attempt at fixing this. Your insights were very valuable, thank you! I hope the query composition is fine, but I assume we'll see in the CI.

File: aries_vcx_core/src/anoncreds/credx_anoncreds.rs
Comment (gmulhearn) [inline aries_vcx_core/src/anoncreds/credx_anoncreds.rs:174]: I think in these cases, we'd still want things to be ANDed together;

I.e. the query is for where 'attr1=="1" && attr2=="2" && attr3=="3" etc'. So I think it's still json({"$and": attrs}).

Either way, these branches are a bit of an edge case. I might make ticket about the wallet queries in this class, we really should be using the Indy-wql crate, or create our own utility for query construction

**Review by Patrik-Stas — APPROVED** [review]

**Review by gmulhearn — APPROVED** [review]
lgtm assuming we make another ticket to address a revision to the wql structures

**Review by mirgee — APPROVED** [review]

**Pull request merged**
